### PR TITLE
core/cmd: port `/R` commands to the rzshell

### PR DESCRIPTION
**Pull request by giridharprasath**
# SQUASH ME

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- Inital Port command yaml changes

ScreenShot_2024-05-24_at_06:36:51-PM

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Review by ret2libc — COMMENTED** [review]
What is the issue related to this? Why under `i` (which is usually for binary-related info)? I think there's something like this already under `/R`.

**Review by XVilka — CHANGES_REQUESTED** [review]
Agree, they should be in the `/R` commands, just put them into:
- `librz/core/cmd_descs/cmd_search.yaml`
- `librz/core/cmd/cmd_search_rop.c` (this one should be updated from the old handlers to use the new handlers.

Moreover, I also recommend adding TABLE output, with more information about each gadget

**Comment by giridharprasath** [comment]
Still WIP:

- Fix memory leaks
- Add table view for gadget
- Regex filter is broken, should be fixed.

File: librz/core/cmd/cmd_search.c [hunk 1]
```diff
@@ -1138,7 +1186,7 @@ static RzList /*<RzCoreAsmHit *>*/ *construct_rop_gadget(RzCore *core, ut64 addr
 	return hitlist;
 }
 
-static void print_rop(RzCore *core, RzList /*<RzCoreAsmHit *>*/ *hitlist, PJ *pj, int mode) {
+static void print_rop(RzCore *core, RzList /*<RzCoreAsmHit *>*/ *hitlist, PJ *pj, RzOutputMode mode) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_search.c]: I recommend passing the whole `RzCmdStateOutput` to this function instead, especially important for adding the TABLE mode.

File: librz/core/cmd/cmd_search.c [hunk 2]
```diff
@@ -246,6 +230,70 @@ static int search_hash(RzCore *core, const char *hashname, const char *hashstr,
...
+				rz_list_free(hitlist);
+				break;
+			}
+			sscanf(s, "%" PFMT64x "(%" PFMT32d ")", &addr, &opsz);
+			hit->addr = addr;
+			hit->len = opsz;
+			rz_list_append(hitlist, hit);
+		} while (*(s = strchr(s, ')') + 1) != '\0');
+
+		print_rop(core, hitlist, pj, state->mode);
```
Comment (XVilka) [inline librz/core/cmd/cmd_search.c]: If you pass `RzCmdStateOutput`, some of the code will not be needed anymore

**Review by XVilka — CHANGES_REQUESTED** [review]
Thanks, much better now!

**Comment by giridharprasath** [comment]
Some notes:
- `/Rk` is unusable, Will be refactoring it and adding testcases for it in a separate PR.
- Current table view with only columns:
```
[0x08048330]> /Rt
addr bytes disasm 
------------------
```
Please lmk if any other fields needed to be added which would be helpful in table mode. @XVilka

**Comment by XVilka** [comment]
> Some notes:
> 
>     * `/Rk` is unusable, Will be refactoring it and adding testcases for it in a separate PR.
> 
>     * Current table view with only columns:
> 
> 
> ```
> [0x08048330]> /Rt
> addr bytes disasm 
> ------------------
> ```
> 
> Please lmk if any other fields needed to be added which would be helpful in table mode. @XVilka

Should be enough for now. Further improvements could be done when necessary.

File: librz/core/cmd/cmd_search.c [hunk 3]
```diff
@@ -1291,39 +1329,124 @@ static void print_rop(RzCore *core, RzList /*<RzCoreAsmHit *>*/ *hitlist, PJ *pj
...
+			char *asmop_hex_str_dup = NULL;
+			if (asmop_hex_str) {
+				asmop_hex_str_dup = rz_asm_op_get_hex(&asmop);
+				asmop_hex_str = rz_str_append(asmop_hex_str, asmop_hex_str_dup);
+			} else {
+				asmop_hex_str_dup = rz_asm_op_get_hex(&asmop);
+				asmop_hex_str = rz_str_newf("%s", asmop_hex_str_dup);
+			}
+			free(asmop_hex_str_dup);
+			// rz_table_add_rowf(state->d.t, "Xss", hit->addr, rz_asm_op_get_hex(&asmop), rz_asm_op_get_asm(&asmop));
```
Comment (XVilka) [inline librz/core/cmd/cmd_search.c]: Why this is commented out?

File: librz/core/cmd/cmd_search.c [hunk 4]
```diff
@@ -1291,39 +1329,124 @@ static void print_rop(RzCore *core, RzList /*<RzCoreAsmHit *>*/ *hitlist, PJ *pj
...
+		default:
+			rz_warn_if_reached();
+			break;
 		}
+	}
+	switch (state->mode) {
+	case RZ_OUTPUT_MODE_JSON:
+		pj_end(state->d.pj);
+		if (db && hit) {
+			// rz_cons_printf ("Gadget size: %d\n", (int)size);
```
Comment (XVilka) [inline librz/core/cmd/cmd_search.c]: Remove commented line

File: librz/core/cmd/cmd_search.c [hunk 5]
```diff
@@ -1291,39 +1329,124 @@ static void print_rop(RzCore *core, RzList /*<RzCoreAsmHit *>*/ *hitlist, PJ *pj
...
+			pj_ki(state->d.pj, "size", size);
+		}
+		pj_end(state->d.pj);
+		break;
+	case RZ_OUTPUT_MODE_QUIET:
+		// fallthrough
+	case RZ_OUTPUT_MODE_STANDARD:
 		if (db && hit) {
-			const ut64 addr = ((RzCoreAsmHit *)rz_list_first(hitlist))->addr;
 			// rz_cons_printf ("Gadget size: %d\n", (int)size);
```
Comment (XVilka) [inline librz/core/cmd/cmd_search.c]: Remove commented line

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks good overall, please address my feedback first though

File: librz/core/cmd/cmd_search.c (hunk 3 above)
Comment (giridharprasath) [reply librz/core/cmd/cmd_search.c]: This needs to be removed, thanks for pointing it out

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
@giridharprasath two more missing failed tests:
```
[XX] db/cmd/cmd_search Rop search w/ regexp
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -qc 'e asm.arch=x86
e analysis.arch=x86
e asm.bits=32
wa "add esp,8;pop ebx; pop ebp; ret"
e scr.color=false
/R/ or.*bl
q
' malloc://1024
-- stdout
--- expected
+++ actual
@@ -0,4 +0,5 @@
   0x00000002             085b5d  or byte [ebx + 0x5d], bl
   0x00000005                 c3  ret
+Gadget size: 4
 

[XX] db/cmd/cmd_search Rop search
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -qc 'e asm.arch=x86
e analysis.arch=x86
e asm.bits=32
wa "add esp,8;pop ebx; pop ebp; ret"
e scr.color=false
/R
q
' malloc://1024
-- stdout
--- expected
+++ actual
@@ -2,12 +2,15 @@
   0x00000003                 5b  pop ebx
   0x00000004                 5d  pop ebp
   0x00000005                 c3  ret
+Gadget size: 6
 ```
https://github.com/rizinorg/rizin/actions/runs/9329582517/job/25682181832?pr=4504#step:12:238

**Pull request merged**